File: apps/driver-android/src/main/kotlin/lk/mageride/driver/notifications/NotificationsScreen.kt
package lk.mageride.driver.notifications

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.driver.R
import lk.mageride.driver.nav.DriverRoute
import lk.mageride.driver.ui.theme.ControlTokens
import lk.mageride.driver.ui.theme.MageRideTheme
import org.koin.androidx.compose.koinViewModel
import kotlin.time.ExperimentalTime

/**
 * **SCR-DA-034 · alerts** (Epic 10, D5' §14.4).
 *
 * The wireframe: a `‹ Alerts` app bar over a list of rows, each a tinted leading square, a headline
 * and a relative time — *"Directional expiring in 10 min · 2 min ago"*.
 *
 * **Shimmer while loading**, which D2' names for this screen: three placeholder rows rather than a
 * spinner, because the list has a shape and a spinner over an empty column reads as an empty inbox.
 *
 * @param onOpen A row's deep link resolved to a screen. Unresolvable links are already `null` — see
 *   [NotificationsViewModel].
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun NotificationsScreen(onBack: () -> Unit, onOpen: (DriverRoute) -> Unit, modifier: Modifier = Modifier) {
    val viewModel: NotificationsViewModel = koinViewModel()
    val state by viewModel.state.collectAsStateWithLifecycle()

    Scaffold(
        modifier = modifier.fillMaxSize(),
        topBar = {
            TopAppBar(
                title = { Text(text = stringResource(R.string.alerts_title)) },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
                            contentDescription = stringResource(R.string.action_back),
                        )
                    }
                },
                actions = {
                    if (state.alerts.any { !it.read }) {
                        TextButton(onClick = viewModel::markAllRead) {
                            Text(text = stringResource(R.string.alerts_mark_all_read))
                        }
                    }
                },
            )
        },
    ) { insets ->
        Column(modifier = Modifier.padding(insets).fillMaxSize()) {
            when {
                state.loading -> AlertShimmer()

                state.isEmpty -> Text(
                    text = stringResource(R.string.alerts_empty),
                    modifier = Modifier.padding(MageRideTheme.spacing.lg),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )

                else -> LazyColumn(
                    modifier = Modifier.fillMaxSize(),
                    contentPadding = PaddingValues(MageRideTheme.spacing.md),
                    verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
                ) {
                    items(state.alerts, key = DriverAlert::id) { alert ->
                        AlertRow(alert = alert, onClick = { viewModel.open(alert) })
                    }
                }
            }
        }
    }

    LaunchedEffect(state.opening) {
        state.opening?.let {
            onOpen(it)
            viewModel.consumeOpening()
        }
    }
}

/** One `.listrow` — the tinted square, the headline and the relative time. */
@OptIn(ExperimentalTime::class)
@Composable
private fun AlertRow(alert: DriverAlert, onClick: () -> Unit, modifier: Modifier = Modifier) {
    val kind = AlertKind.of(alert.type)

    Row(
        modifier = modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(vertical = MageRideTheme.spacing.xs),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
    ) {
        Box(
            modifier = Modifier
                .size(ControlTokens.ListRowIcon)
                .background(
                    color = alertTone(kind).copy(alpha = SQUARE_TINT),
                    shape = RoundedCornerShape(MageRideTheme.radius.sm),
                ),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = kind.icon,
                contentDescription = null,
                modifier = Modifier.size(ControlTokens.ChipIcon),
                tint = alertTone(kind),
            )
        }

        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = alert.title ?: stringResource(kind.label),
                style = MaterialTheme.typography.bodyMedium,
                // An unread alert is the one thing on this list a driver is scanning for.
                fontWeight = if (alert.read) FontWeight.Normal else FontWeight.SemiBold,
                color = MaterialTheme.colorScheme.onSurface,
            )
            alert.body?.takeIf(String::isNotBlank)?.let { body ->
                Text(
                    text = body,
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Text(
                text = alertAge(alert.receivedAt),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.outlineVariant,
            )
        }
    }
}

/** D2' §SCR-DA-034's *"shimmer loading"* — three rows of the shape that is coming. */
@Composable
private fun AlertShimmer(modifier: Modifier = Modifier) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(MageRideTheme.spacing.md),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
    ) {
        repeat(SHIMMER_ROWS) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
            ) {
                Box(
                    modifier = Modifier
                        .size(ControlTokens.ListRowIcon)
                        .background(
                            MaterialTheme.colorScheme.surfaceVariant,
                            RoundedCornerShape(MageRideTheme.radius.sm),
                        ),
                )
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .height(ControlTokens.ListRowIcon)
                        .background(
                            MaterialTheme.colorScheme.surfaceVariant,
                            RoundedCornerShape(MageRideTheme.radius.sm),
                        ),
                )
            }
        }
    }
}

/** How much of the tone tints a row's leading square. Light enough to read the icon over. */
private const val SQUARE_TINT = 0.18f

/** The wireframe draws four rows; three is enough to say "a list is coming". */
private const val SHIMMER_ROWS = 3

File: apps/driver-ios/DriverApp/Notifications/NotificationsScreen.swift
import SwiftUI

/// **SCR-DI-034 · alerts** (Epic 10, D5' §14.4).
///
/// The wireframe: an `Alerts` large title over a grouped list, each row a tinted leading square, a
/// headline and a relative time — *"Directional expiring in 10 min · 2 min ago"*.
///
/// **Shimmer while loading**, which D2' names for this screen: four placeholder rows rather than a
/// spinner, because the list has a shape and a spinner over an empty column reads as an empty inbox.
/// The shimmer is a `.redacted(reason: .placeholder)` group — the platform's own version of the
/// effect, so it respects Reduce Motion without a branch here.
///
/// It hangs off the **Menu** tab, so the system back button says `‹ Menu`.
///
/// - Parameter onOpen: A row's deep link resolved to a screen. Unresolvable links are already `nil`
///   — see ``NotificationsModel``.
@MainActor
struct NotificationsScreen: View {

    @StateObject private var model: NotificationsModel

    private let onOpen: (DriverRoute) -> Void

    init(model: @autoclosure @escaping () -> NotificationsModel, onOpen: @escaping (DriverRoute) -> Void) {
        _model = StateObject(wrappedValue: model())
        self.onOpen = onOpen
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.xs) {
                if model.state.isLoading {
                    shimmer
                } else if model.state.isEmpty {
                    Text(key: "alerts_empty")
                        .mageFont(.body)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                        .padding(.top, MageRideSpacing.lg)
                } else {
                    // ``DriverAlert`` is `Identifiable` on §1.6's own primary key, so `ForEach` needs
                    // no `id:` — and it must not be given one built from `Array(_:enumerated())`,
                    // because **Swift has no key paths into tuples** (the C087 finding).
                    GroupedList {
                        ForEach(model.state.alerts) { alert in
                            row(alert, showsSeparator: alert.id != model.state.alerts.last?.id)
                        }
                    }
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.surface)
        .navigationTitle(Text(key: "alerts_title"))
        .navigationBarTitleDisplayMode(.large)
        .toolbar {
            if model.state.hasUnread {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button(action: model.markAllRead) {
                        Text(key: "alerts_mark_all_read")
                    }
                }
            }
        }
        .task { await model.refresh() }
        .onChange(of: model.state.opening) { route in
            guard let route else { return }
            model.consumeOpening()
            onOpen(route)
        }
    }

    /// One `.glist .gr` — the tinted square, the headline and the relative time.
    private func row(_ alert: DriverAlert, showsSeparator: Bool) -> some View {
        let kind = AlertKind.of(alert.type)

        return VStack(spacing: 0) {
            Button { model.open(alert) } label: {
                HStack(spacing: MageRideSpacing.sm) {
                    Image(systemName: kind.symbolName)
                        .font(.footnote)
                        .foregroundStyle(kind.accent)
                        .frame(width: MageRideControl.listRowIcon, height: MageRideControl.listRowIcon)
                        .background(
                            kind.accent.opacity(Self.squareTint),
                            in: RoundedRectangle(cornerRadius: MageRideRadius.sm, style: .continuous)
                        )

                    VStack(alignment: .leading, spacing: 1) {
                        Text(alert.title ?? kind.labelKey.localised)
                            // An unread alert is the one thing on this list a driver is scanning for.
                            .mageFont(alert.isRead ? .body : .bodyEmphasis)
                            .foregroundStyle(MageRideColor.onSurface)
                            .multilineTextAlignment(.leading)

                        if let body = alert.body, !body.isEmpty {
                            Text(body)
                                .mageFont(.caption)
                                .foregroundStyle(MageRideColor.onSurfaceVariant)
                                .multilineTextAlignment(.leading)
                        }

                        Text(AlertAge.of(receivedAt: alert.receivedAt, now: Date()).text)
                            .mageFont(.caption)
                            .foregroundStyle(MageRideColor.outlineVariant)
                    }

                    Spacer(minLength: MageRideSpacing.xs)
                }
                .padding(.horizontal, MageRideSpacing.sm)
                .padding(.vertical, MageRideSpacing.xs)
                .frame(minHeight: MageRideControl.minimumTapTarget)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityElement(children: .combine)

            if showsSeparator {
                Rectangle()
                    .fill(MageRideColor.surfaceVariant)
                    .frame(height: MageRideControl.hairline)
                    .padding(.leading, MageRideSpacing.sm + MageRideControl.listRowIcon + MageRideSpacing.sm)
            }
        }
    }

    /// D2' §SCR-DI-034's *"shimmer loading"* — the shape that is coming, redacted.
    private var shimmer: some View {
        GroupedList {
            ForEach(0..<Self.shimmerRows, id: \.self) { index in
                row(
                    DriverAlert(
                        id: "placeholder-\(index)",
                        type: "UNKNOWN",
                        title: String(repeating: " ", count: 24),
                        receivedAt: Date()
                    ),
                    showsSeparator: index < Self.shimmerRows - 1
                )
            }
        }
        .redacted(reason: .placeholder)
        .disabled(true)
        .accessibilityHidden(true)
    }

    /// How much of the tone tints a row's leading square. Light enough to read the icon over.
    private static let squareTint: Double = 0.18

    /// The wireframe draws four rows, and so does this.
    private static let shimmerRows = 4
}
